katsdptelstate/memory.py

```diff
@@ -144,99 +146,130 @@ class MemoryBackend(Backend):
     Write operations are supported only to facilitate testing, but are not
     intended for production use. For that, use a
     :class:`~katsdptelstate.redis.RedisBackend` with an in-memory Redis
-    emulation. The :meth:`monitor_keys` method is not implemented.
+    emulation.
 
     Mutable keys are stored as sorted lists, and encode timestamps in-place
     using the same packing as :class:`~katsdptelstate.redis.RedisBackend`.
     """
     def __init__(self) -> None:
         self._data = {}       # type: Dict[bytes, _Value]
+        self._generation = 0  # incremented each time a change is made
+        self._condition = threading.Condition()
+
+    @contextlib.contextmanager
+    def _write(self) -> Generator[None, None, None]:
+        """Context manager to use for writes."""
+        with self._condition:
+            yield
+            self._generation += 1
```

Comment:
I don't really follow the need for _generation to identify changes, since the condition.wait_for() is only triggered when _generation is incremented, why not just condition.wait() and not bother with the _generation at all ?

Reply:
It's needed in the async MemoryBackend, which starts a task on a worker thread to wait for updates. Without recording the generation, there would be a race condition where the database was updated after submitting/starting the task but before that task could obtain the lock. The async MemoryBackend also signals the condition variable without a database change to unblock that worker thread.